syntax = "proto3";

package rds;

/*
If you need to add some proto files for generic usage,
Not related to grpc server or client,

Its' recommended to keep those in this file.
*/

enum DBType {
  UNKNOWN_DB_TYPE = 0;
  MYSQL = 1;
  MARIADB = 2;
}

message DBHealthStatus {
  DBType db_type = 1;
  int64 reported_at = 2; // Timestamp in milliseconds
  string global_transaction_id = 3; // Useful for MySQL/MariaDB
}

enum ClusterNodeType {
  UNKNOWN_CLUSTER_NODE_TYPE = 0;
  MASTER = 1;
  REPLICA = 2;
  READ_ONLY = 3;
  STANDBY = 4;
}

enum ClusterNodeStatus {
  UNKNOWN_CLUSTER_NODE_STATUS = 0;
  ONLINE = 1;
  OFFLINE = 2;
  MAINTENANCE = 3;
}

message ClusterNodeConfig {
  ClusterNodeType type = 1;
  ClusterNodeStatus status = 2;
  string ip = 3;
  int32 agent_port = 4;
  int32 db_port = 5;
  int32 weight = 6;
}

message ClusterProxyConfig {
  string ip = 1;
  int32 agent_port = 2;
}

message ClusterConfig {
  int32 max_replication_lag_ms = 1;
  int32 failover_cooldown_ms = 2;
  int64 last_failover_timestamp = 3;
  map<string, ClusterNodeConfig> nodes = 4;
  optional ClusterProxyConfig proxy = 5;
  string replication_user = 6;
  string replication_password = 7;
  string shared_token = 8;
}